### PR TITLE
bpo-35081: Enhance Python-ast.h and ast.h

### PR DESCRIPTION
**Pull request by vstinner**
Modify asdl_c.py to enhance Python-ast.h:

* Add #ifndef/#define PYTHON_AST_H to be able to include the header
  twice
* Add "extern { ... }" for C++
* Undefine "Yield" macro conflicting with winbase.h
* ast.h: add node.h and Python-ast.h includes

https://bugs.python.org/issue35081

### CODE REVIEW

**Comment by serhiy-storchaka** [comment]
What problem does it solve? I have not found anything on bpo-35081, may be open a new issue for discussing?

**Comment by vstinner** [comment]
> What problem does it solve?

I wanted to reorder includes in pylifecycle.c, but ast.h doesn't declare properly its dependencies: it uses node which comes from node.h, and mod_ty which comes from Python-ast.h. This PR fix this.

Undefining the "Yield" macro fix a warning on Windows. In practice, it's more a cleanup.

> I have not found anything on bpo-35081, may be open a new issue for discussing?

Well, it's more a "cleanup" change, so I didn't want to open a new issue just for that.

**Review by ncoghlan — APPROVED** [review]
LGTM, but a comment should be added to bpo-35081 noting the change to make ast.h easier to include correctly.

File: Include/Python-ast.h [hunk 1]
```diff
@@ -1,7 +1,15 @@
 /* File automatically generated by Parser/asdl_c.py. */
 
+#ifndef PYTHON_AST_H
+#define PYTHON_AST_H
+#ifdef __cplusplus
+extern "C" {
```
Comment (serhiy-storchaka) [inline Include/Python-ast.h:6]: Do we need this? This is an internal generated header, it is not purposed for direct use.

File: Include/ast.h [hunk 2]
```diff
@@ -4,6 +4,9 @@
 extern "C" {
 #endif
 
+#include "node.h"         /* node */
```
Comment (serhiy-storchaka) [inline Include/ast.h:7]: Other option is to add forward declarations, as in `compile.h`.

File: Include/Python-ast.h (hunk 1 above)
Comment (vstinner) [reply Include/Python-ast.h:6]: Yes, it's needed if ast.h is included in file compiled in C++. I know that some people do that. At least, it doesn't hurt to add extern "C" { ... }.

File: Include/ast.h (hunk 2 above)
Comment (vstinner) [reply Include/ast.h:7]: Well, maybe, but it seems like node.h is included anyway in all C files including ast.h. I don't see the benefit of avoiding to include node.h here. I prefer to be explicit and link ast.h to node.h.

Comment (serhiy-storchaka) [reply Include/ast.h:7]: `#include "Python-ast.h"` is already surrounded with `extern "C" {` in ast.h.

**Comment by vstinner** [comment]
Superseded by https://bugs.python.org/issue35177 and PR #10361.

**Pull request closed without merging**